panic: close of closed channel

Not sure how this might happen but I got:

```
panic: close of closed channel

goroutine 15 [running]:
github.com/wagslane/go-rabbitmq.(*channelManager).startNotifyCancelOrClosed(0xc0000c4840)
        /var/lib/jenkins/go/pkg/mod/github.com/wagslane/go-rabbitmq@v0.6.2-0.20210618171750-0f20affc2859/channel.go:85 +0x1c5
created by github.com/wagslane/go-rabbitmq.newChannelManager
        /var/lib/jenkins/go/pkg/mod/github.com/wagslane/go-rabbitmq@v0.6.2-0.20210618171750-0f20affc2859/channel.go:35 +0x21a
```

It is probably happening because channel was closed inside NotifyCancel method https://github.com/wagslane/go-rabbitmq/blob/0f20affc2859ccb8250fd0380994c324ef2c287f/channel.go#L59

--- a/channel.go
+++ b/channel.go
@@ -53,10 +53,9 @@ func getNewChannel(url string, conf amqp.Config) (*amqp.Connection, *amqp.Channe
 // backoff. Once reconnected, it sends an error back on the manager's notifyCancelOrClose
 // channel
 func (chManager *channelManager) startNotifyCancelOrClosed() {
-	notifyCloseChan := make(chan *amqp.Error)
-	notifyCloseChan = chManager.channel.NotifyClose(notifyCloseChan)
-	notifyCancelChan := make(chan string)
-	notifyCancelChan = chManager.channel.NotifyCancel(notifyCancelChan)
+	notifyCloseChan := chManager.channel.NotifyClose(make(chan *amqp.Error, 1))
+	notifyCancelChan := chManager.channel.NotifyCancel(make(chan string, 1))
+
 	select {
 	case err := <-notifyCloseChan:
 		// If the connection close is triggered by the Server, a reconnection takes place
@@ -71,18 +70,6 @@ func (chManager *channelManager) startNotifyCancelOrClosed() {
 		chManager.reconnectWithBackoff()
 		chManager.logger.Printf("successfully reconnected to amqp server after cancel")
 		chManager.notifyCancelOrClose <- errors.New(err)
-	}
-
-	// these channels can be closed by amqp
-	select {
-	case <-notifyCloseChan:
-	default:
-		close(notifyCloseChan)
-	}
-	select {
-	case <-notifyCancelChan:
-	default:
-		close(notifyCancelChan)
 	}
 }
 

--- a/publish.go
+++ b/publish.go
@@ -99,6 +99,8 @@ func WithPublishOptionsHeaders(headers Table) func(*PublishOptions) {
 type Publisher struct {
 	chManager *channelManager
 
+	notifyReturnChan chan Return
+
 	disablePublishDueToFlow    bool
 	disablePublishDueToFlowMux *sync.RWMutex
 
@@ -148,21 +150,21 @@ func NewPublisher(url string, config amqp.Config, optionFuncs ...func(*Publisher
 
 	publisher := Publisher{
 		chManager:                  chManager,
+		notifyReturnChan:           make(chan Return, 1),
 		disablePublishDueToFlow:    false,
 		disablePublishDueToFlowMux: &sync.RWMutex{},
 		logger:                     options.Logger,
 	}
 
-	returnChan := make(chan Return)
 	go func() {
-		publisher.startNotifyHandlers(returnChan)
+		publisher.startNotifyHandlers()
 		for err := range publisher.chManager.notifyCancelOrClose {
 			publisher.logger.Printf("publish cancel/close handler triggered. err: %v", err)
-			publisher.startNotifyHandlers(returnChan)
+			publisher.startNotifyHandlers()
 		}
 	}()
 
-	return publisher, returnChan, nil
+	return publisher, publisher.notifyReturnChan, nil
 }
 
 // Publish publishes the provided data to the given routing keys over the connection
@@ -215,11 +217,11 @@ func (publisher Publisher) StopPublishing() {
 	publisher.chManager.connection.Close()
 }
 
-func (publisher *Publisher) startNotifyHandlers(returnChan chan Return) {
+func (publisher *Publisher) startNotifyHandlers() {
 	returnAMQPChan := publisher.chManager.channel.NotifyReturn(make(chan amqp.Return, 1))
 	go func() {
 		for ret := range returnAMQPChan {
-			returnChan <- Return{ret}
+			publisher.notifyReturnChan <- Return{ret}
 		}
 	}()
 

--- a/publish.go
+++ b/publish.go
@@ -155,22 +155,10 @@ func NewPublisher(url string, config amqp.Config, optionFuncs ...func(*Publisher
 
 	returnChan := make(chan Return)
 	go func() {
-		setupNotifyChans := func() {
-			returnAMQPChan := publisher.chManager.channel.NotifyReturn(make(chan amqp.Return))
-			go func() {
-				for ret := range returnAMQPChan {
-					returnChan <- Return{ret}
-				}
-			}()
-
-			notifyFlowChan := publisher.chManager.channel.NotifyFlow(make(chan bool))
-			go publisher.startNotifyFlowHandler(notifyFlowChan)
-		}
-
-		setupNotifyChans()
+		publisher.startNotifyHandlers(returnChan)
 		for err := range publisher.chManager.notifyCancelOrClose {
 			publisher.logger.Printf("publish cancel/close handler triggered. err: %v", err)
-			setupNotifyChans()
+			publisher.startNotifyHandlers(returnChan)
 		}
 	}()
 
@@ -225,6 +213,18 @@ func (publisher *Publisher) Publish(
 func (publisher Publisher) StopPublishing() {
 	publisher.chManager.channel.Close()
 	publisher.chManager.connection.Close()
+}
+
+func (publisher *Publisher) startNotifyHandlers(returnChan chan Return) {
+	returnAMQPChan := publisher.chManager.channel.NotifyReturn(make(chan amqp.Return, 1))
+	go func() {
+		for ret := range returnAMQPChan {
+			returnChan <- Return{ret}
+		}
+	}()
+
+	notifyFlowChan := publisher.chManager.channel.NotifyFlow(make(chan bool))
+	go publisher.startNotifyFlowHandler(notifyFlowChan)
 }
 
 func (publisher *Publisher) startNotifyFlowHandler(notifyFlowChan chan bool) {

--- a/publish.go
+++ b/publish.go
@@ -230,6 +230,10 @@ func (publisher *Publisher) startNotifyHandlers() {
 }
 
 func (publisher *Publisher) startNotifyFlowHandler(notifyFlowChan chan bool) {
+	publisher.disablePublishDueToFlowMux.Lock()
+	publisher.disablePublishDueToFlow = false
+	publisher.disablePublishDueToFlowMux.Unlock()
+
 	// Listeners for active=true flow control.  When true is sent to a listener,
 	// publishing should pause until false is sent to listeners.
 	for ok := range notifyFlowChan {

--- a/publish.go
+++ b/publish.go
@@ -99,8 +99,6 @@ func WithPublishOptionsHeaders(headers Table) func(*PublishOptions) {
 type Publisher struct {
 	chManager *channelManager
 
-	notifyFlowChan chan bool
-
 	disablePublishDueToFlow    bool
 	disablePublishDueToFlowMux *sync.RWMutex
 
@@ -150,26 +148,31 @@ func NewPublisher(url string, config amqp.Config, optionFuncs ...func(*Publisher
 
 	publisher := Publisher{
 		chManager:                  chManager,
-		notifyFlowChan:             make(chan bool),
 		disablePublishDueToFlow:    false,
 		disablePublishDueToFlowMux: &sync.RWMutex{},
 		logger:                     options.Logger,
 	}
 
-	returnAMQPChan := make(chan amqp.Return)
 	returnChan := make(chan Return)
-	returnAMQPChan = publisher.chManager.channel.NotifyReturn(returnAMQPChan)
 	go func() {
-		for ret := range returnAMQPChan {
-			returnChan <- Return{
-				ret,
-			}
+		setupNotifyChans := func() {
+			returnAMQPChan := publisher.chManager.channel.NotifyReturn(make(chan amqp.Return))
+			go func() {
+				for ret := range returnAMQPChan {
+					returnChan <- Return{ret}
+				}
+			}()
+
+			notifyFlowChan := publisher.chManager.channel.NotifyFlow(make(chan bool))
+			go publisher.startNotifyFlowHandler(notifyFlowChan)
+		}
+
+		setupNotifyChans()
+		for err := range publisher.chManager.notifyCancelOrClose {
+			publisher.logger.Printf("publish cancel/close handler triggered. err: %v", err)
+			setupNotifyChans()
 		}
 	}()
-
-	publisher.notifyFlowChan = publisher.chManager.channel.NotifyFlow(publisher.notifyFlowChan)
-
-	go publisher.startNotifyFlowHandler()
 
 	return publisher, returnChan, nil
 }
@@ -224,10 +227,10 @@ func (publisher Publisher) StopPublishing() {
 	publisher.chManager.connection.Close()
 }
 
-func (publisher *Publisher) startNotifyFlowHandler() {
+func (publisher *Publisher) startNotifyFlowHandler(notifyFlowChan chan bool) {
 	// Listeners for active=true flow control.  When true is sent to a listener,
 	// publishing should pause until false is sent to listeners.
-	for ok := range publisher.notifyFlowChan {
+	for ok := range notifyFlowChan {
 		publisher.disablePublishDueToFlowMux.Lock()
 		if ok {
 			publisher.logger.Printf("pausing publishing due to flow request from server")

